src/platform_impl/web/event_loop/proxy.rs

```diff
@@ -1,26 +1,96 @@
-use super::runner;
-use crate::event::Event;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::mpsc::{self, Receiver, RecvError, SendError, Sender, TryRecvError};
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Waker};
+
 use crate::event_loop::EventLoopClosed;
 
 pub struct EventLoopProxy<T: 'static> {
-    runner: runner::Shared<T>,
+    sender: AsyncSender<T>,
 }
 
 impl<T: 'static> EventLoopProxy<T> {
-    pub fn new(runner: runner::Shared<T>) -> Self {
-        Self { runner }
+    pub fn new(sender: AsyncSender<T>) -> Self {
+        Self { sender }
     }
 
     pub fn send_event(&self, event: T) -> Result<(), EventLoopClosed<T>> {
-        self.runner.send_event(Event::UserEvent(event));
-        Ok(())
+        match self.sender.send(event) {
+            Ok(()) => Ok(()),
+            Err(SendError(val)) => Err(EventLoopClosed(val)),
+        }
     }
 }
 
 impl<T: 'static> Clone for EventLoopProxy<T> {
     fn clone(&self) -> Self {
         Self {
-            runner: self.runner.clone(),
+            sender: self.sender.clone(),
+        }
+    }
+}
+
+pub fn channel<T: 'static>() -> (AsyncSender<T>, AsyncReceiver<T>) {
+    let (sender, receiver) = mpsc::channel();
+    let waker = Arc::new(Mutex::new(None));
+
+    let sender = AsyncSender {
+        sender,
+        waker: Arc::clone(&waker),
+    };
+    let receiver = AsyncReceiver { receiver, waker };
+
+    (sender, receiver)
+}
+
+pub struct AsyncSender<T: 'static> {
+    sender: Sender<T>,
+    waker: Arc<Mutex<Option<Waker>>>,
+}
+
+impl<T: 'static> AsyncSender<T> {
+    pub fn send(&self, event: T) -> Result<(), SendError<T>> {
+        self.sender.send(event)?;
+
+        if let Some(waker) = self.waker.lock().unwrap().take() {
+            waker.wake();
+        }
+
+        Ok(())
+    }
+}
+
+impl<T: 'static> Clone for AsyncSender<T> {
+    fn clone(&self) -> Self {
+        Self {
+            sender: self.sender.clone(),
+            waker: self.waker.clone(),
+        }
+    }
+}
+
+pub struct AsyncReceiver<T: 'static> {
+    receiver: Receiver<T>,
+    waker: Arc<Mutex<Option<Waker>>>,
+}
+
+impl<T: 'static> Future for AsyncReceiver<T> {
+    type Output = Result<T, RecvError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.receiver.try_recv() {
+            Ok(event) => Poll::Ready(Ok(event)),
+            Err(TryRecvError::Empty) => {
+                *self.waker.lock().unwrap() = Some(cx.waker().clone());
```

Comment:
You should use `will_wake` here in order to avoid unnecessarily cloning a waker twice.

Reply:
I didn't know about this! I also noticed that neither `futures` or `atomic-waker` use this this optimization in their `AtomicWaker` implementation. I open an issue here: https://github.com/smol-rs/atomic-waker/issues/11.

In our case this won't be possible because to avoid a racing condition we would have to hold the lock, which can lead to blocking in a multi-threaded environment.

Reply:
Couldn't you do:

```rust
{
    let mut waker = self.waker.lock().unwrap();
    if waker.map_or(true, |waker| !waker.will_wake(cx.waker())) {
        *waker = Some(cx.waker().clone());
    }
}
```

in order to drop the lock once the temporary scope ends?

Reply:
I wasn't concerned with how to drop the lock, I was concerned with the fact that we hold it in the first place. The assumption in play here is that if you only assign a value without any checks in-between, optimizations can prevent this to block the main thread. Not really a provable assumption, but it's what we have right now without using `AtomicWaker`.

If the the worker yields between holding the lock and comparing and cloning the `Waker`, we are gonna have a problem.